### PR TITLE
Dt/enhancement regulator timeout

### PR DESCRIPTION
**Pull request by danthompson41**
71ij719sfzl _sl1094_

This adds a timeout to the regulator. If the proper pressure isn't held for longer than 5 seconds, the pressure target will be set to 0 and an error will be thrown. If it was too low, it'll be marked as a leak and if it was too high it'll be marked as a runaway condition. Acceptable value band size suggested by Travis, easily modifiable. 

- [x] Unit Test
- [x] Alignment
- [x] Measure Ram - 969 bytes at startup

### CODE REVIEW

**Comment by danthompson41** [comment]
@kdumontnu @pizzyflavin @jminardi

File: Marlin/Regulator.cpp [hunk 1]
```diff
@@ -9,23 +9,136 @@
...
+
+ /** 
+  * Allows us to determine if the pressure is too high or low, indicating
+  * a serious issue. Checks the target temperature vs. the desired, and 
+  * throws an error if outside the bounds for too long.
+  * @target_pressure  The pressure you're attempting to reach
+  * @pressure         The actual current pressure 
+  */
+static void pressure_protection(float target_pressure, float pressure) {
+  static const int period_seconds = REGULATOR_PROTECTION_TIME_S;
```
Comment (kdumontnu) [inline Marlin/Regulator.cpp]: Is this used anywhere else?

File: Marlin/Regulator.cpp [hunk 2]
```diff
@@ -9,23 +9,136 @@
...
+
+// Relevant Pins
+// Voxel8 C2
+// #define PNEUMATIC_PUMP_PIN    7   // Pump driver
+// #define PNEUMATIC_PIN         5   // A5 Input analog tank pressure reading
+// #define PNEUMATIC_CART1_PIN   4   // A4 Cartridge 2 analog pressure reading
+
+#define REGULATOR_PROTECTION_TIME_S (5)
+#define REGULATOR_PROTECTION_BAND   (4)
+
```
Comment (kdumontnu) [inline Marlin/Regulator.cpp]: Should these defines move to the header file?

**Comment by kdumontnu** [comment]
@danthompson41 I left a couple comments. Once you address those, it looks good.

Has this been tested through a full alignment and test print?

File: Marlin/Regulator.cpp (hunk 1 above)
Comment (danthompson41) [inline Marlin/Regulator.cpp]: Na, I get where you're going with it. The code could just be in the update function. I like keeping it separate like this in case we're going to add anything else to the update, but it could be moved up

Comment (kdumontnu) [inline Marlin/Regulator.cpp]: Nah, I just meant `period_seconds `

File: tests/console.log [hunk 3]
```diff
@@ -1,421 +1,61 @@
-16-05-16 17:15:39 DEBUG    Logging and Parser Complete
```
Comment (kdumontnu) [inline tests/console.log]: Can this file be removed and added to .gitignore?

File: Marlin/temperature.cpp [hunk 4]
```diff
@@ -689,6 +689,7 @@ void manage_heater() {
   if (!temp_meas_ready) return;
 
   updateTemperaturesFromRawValues();
+  Regulator__Update();
```
Comment (kdumontnu) [inline Marlin/temperature.cpp]: I think this should be added under REGULATOR_CHECK_INTERVAL. We don't great time resolution on it. I think REGULATOR_CHECK_INTERVAL should be reduced to ~1s (from 2.5s). @pizzyflavin thoughts?

File: tests/console.log (hunk 3 above)
Comment (danthompson41) [inline tests/console.log]: Done

File: Marlin/temperature.cpp (hunk 4 above)
Comment (danthompson41) [inline Marlin/temperature.cpp]: I didn't actually realize that existed haha moving it there no problem.

File: Marlin/Regulator.h [hunk 5]
```diff
@@ -33,9 +40,25 @@
 #endif // E_REGULATOR_SENSOR
 #define REGULATOR_LOW_P 2
 
+#define REGULATOR_PROTECTION_TIME_S (5)
+#define REGULATOR_PROTECTION_BAND   (4)
+
+#define REGULATOR_NOT_PRESENT_VALUE (130.5)
```
Comment (kdumontnu) [inline Marlin/Regulator.h]: @danthompson41 have you tried this on a few printers? I feel like this isn't consistent, but I could be wrong. Is this just the max-value that the ADC reports?

File: .gitignore [hunk 6]
```diff
@@ -31,3 +31,4 @@ src
 *.pyc
 .build_log
 *.hex
+tests/console.log
```
Comment (pizzyflavin) [inline .gitignore]: Add a newline

File: Marlin/Marlin_main.cpp [hunk 7]
```diff
@@ -5130,9 +5130,9 @@ inline void gcode_M252() {
         servo[servo_index].move(servo_position);
       else {
         SERIAL_ECHO_START;
-        SERIAL_ECHO("Servo ");
+        SERIAL_PROTOCOLPGM("Servo ");
```
Comment (pizzyflavin) [inline Marlin/Marlin_main.cpp]: These are unrelated to the regulator, just putting these strings in prog mem right?

File: Marlin/Regulator.cpp [hunk 8]
```diff
@@ -9,9 +9,45 @@
...
+//===========================================================================
+//=============================== Definitions ===============================
+//===========================================================================
+
+// Relevant Pins
+// Voxel8 C2
+// #define PNEUMATIC_PUMP_PIN    7   // Pump driver
+// #define PNEUMATIC_PIN         5   // A5 Input analog tank pressure reading
+// #define PNEUMATIC_CART1_PIN   4   // A4 Cartridge 2 analog pressure reading
+
```
Comment (pizzyflavin) [inline Marlin/Regulator.cpp]: These should be defined in the pin mapping file

File: Marlin/Regulator.cpp [hunk 9]
```diff
@@ -33,4 +69,98 @@ void setOutputPressure(float desired_pressure) {
...
+//===========================================================================
+
+/** 
+ * Error handler when marlin detects a missing pressure regulator
+ */
+static void _regulator_sled_removed_error() {
+  if (sled_present == true) {
+    //quickStop();
+    //disable_all_heaters();
+    //disable_all_steppers();
```
Comment (pizzyflavin) [inline Marlin/Regulator.cpp]: dead code

File: Marlin/Regulator.cpp [hunk 10]
```diff
@@ -9,9 +9,45 @@
...
+static void _regulator_runaway_error();
+static void pressure_protection(float pressure, float target_pressure);
+
+//===========================================================================
+//============================= Public Functions ============================
+//===========================================================================
+
+/**
+ * Set the output pressure (in psi) of the regulator.
+ * @desired_pressure  The pressure that the function is attempting to
```
Comment (pizzyflavin) [inline Marlin/Regulator.cpp]: @param ?

File: Marlin/Regulator.cpp [hunk 11]
```diff
@@ -9,9 +9,45 @@
...
+// #define PNEUMATIC_PUMP_PIN    7   // Pump driver
+// #define PNEUMATIC_PIN         5   // A5 Input analog tank pressure reading
+// #define PNEUMATIC_CART1_PIN   4   // A4 Cartridge 2 analog pressure reading
+
+//===========================================================================
+//============================ Private Variables ============================
+//===========================================================================
+
+static float current_target_pressure = 0;
+static bool  sled_present = true;
```
Comment (pizzyflavin) [inline Marlin/Regulator.cpp]: It feels unsafe to assume the sled is initially present

File: Marlin/Regulator.cpp [hunk 12]
```diff
@@ -33,4 +69,98 @@ void setOutputPressure(float desired_pressure) {
...
+
+//===========================================================================
+//============================ Private Functions ============================
+//===========================================================================
+
+/** 
+ * Error handler when marlin detects a missing pressure regulator
+ */
+static void _regulator_sled_removed_error() {
+  if (sled_present == true) {
```
Comment (pizzyflavin) [inline Marlin/Regulator.cpp]: It seems to me that the checking of `sled_present` should be done outside of the error function. If this function is ever called, it should execute the body.

File: Marlin/temperature.cpp [hunk 13]
```diff
@@ -787,10 +787,11 @@ void manage_heater() {
   // ELECTRO-PNEUMATIC REGULATOR CONTROL
   #if (ENABLED(E_REGULATOR) && ENABLED(PNEUMATICS))
   if (millis() - previous_millis_regulator_value > REGULATOR_CHECK_INTERVAL) {
-    
+    // Updates the regulator protection timers
+    Regulator__Update();
     previous_millis_regulator_value = millis();
     //Is output pressure more than what is available?
-    if((pressureRegulator() >= pressurePneumatic()) && (pressurePneumatic() >= 1)) {
+    if((pressureRegulator() >= pressurePneumatic()) && (pressurePneumatic() >= REGULATOR_LOW_P) && (pressureRegulator() != REGULATOR_NOT_PRESENT_VALUE)) {
```
Comment (kdumontnu) [inline Marlin/temperature.cpp]: We should consolidate these two different checks we are doing, and handle them the same way.

File: Marlin/Regulator.cpp [hunk 14]
```diff
@@ -33,4 +69,98 @@ void setOutputPressure(float desired_pressure) {
...
+    //quickStop();
+    //disable_all_heaters();
+    //disable_all_steppers();
+    SERIAL_PROTOCOLLNPGM(  
+      "Sled not detected. Canceling print, check pneumatics sled.");
```
Comment (pizzyflavin) [inline Marlin/Regulator.cpp]: "Pneumatics not detected" might be a more helpful message than "Sled not detected" for the initial sentence

File: Marlin/Regulator.cpp [hunk 15]
```diff
@@ -33,4 +69,98 @@ void setOutputPressure(float desired_pressure) {
...
+  static millis_t regulatorTimer = 0;
+
+  if (regulatorTimer == 0) regulatorTimer = millis();
+  bool regulatorUnder =
+      (pressure <= (target_pressure - REGULATOR_PROTECTION_BAND));
+  bool regulatorOver =
+      (pressure >= (target_pressure + REGULATOR_PROTECTION_BAND));
+
+  if (regulatorOver || regulatorUnder) {
+    if (millis() > (regulatorTimer + REGULATOR_PROTECTION_TIME_S * 1000UL)) {
```
Comment (kdumontnu) [inline Marlin/Regulator.cpp]: This is probably optimized out anyways, but should you just make REGULATOR_PROTECTION_TIME_MS

File: Marlin/Regulator.cpp [hunk 16]
```diff
@@ -33,4 +69,98 @@ void setOutputPressure(float desired_pressure) {
...
+ * Error handler when marlin detects a regulator leak 
+ */
+static void _regulator_leak_error() {
+  quickStop();
+  disable_all_heaters();
+  disable_all_steppers();
+  SERIAL_PROTOCOLLNPGM(  
+      "Pressure Leak Error. Canceling print, check the pressure hose");
+  SERIAL_PROTOCOLLNPGM("// action:cancel");
+  Regulator__SetOutputPressure(0.0);
```
Comment (pizzyflavin) [inline Marlin/Regulator.cpp]: Maybe set this pressure to zero psi before the SERIAL stuff. Seems more important to actually turn off the pressure than to print the action.

File: Marlin/Regulator.cpp [hunk 17]
```diff
@@ -33,4 +69,98 @@ void setOutputPressure(float desired_pressure) {
...
+  sled_present = true;
```
Comment (pizzyflavin) [inline Marlin/Regulator.cpp]: ?

File: Marlin/Regulator.cpp [hunk 18]
```diff
@@ -33,4 +69,98 @@ void setOutputPressure(float desired_pressure) {
...
+ * Error handler when marlin detects a regulator pressure that's too high
+ */
+static void _regulator_runaway_error() {
+  quickStop();
+  disable_all_heaters();
+  disable_all_steppers();
+  SERIAL_PROTOCOLLNPGM("Runaway Pressure Error. Canceling print.");
+  SERIAL_PROTOCOLLNPGM("// action:cancel");
+  Regulator__SetOutputPressure(0.0);
+  sled_present = true;
```
Comment (pizzyflavin) [inline Marlin/Regulator.cpp]: still ?

File: Marlin/Regulator.cpp [hunk 19]
```diff
@@ -33,4 +69,98 @@ void setOutputPressure(float desired_pressure) {
...
+/** 
+ * Error handler when marlin detects a regulator pressure that's too high
+ */
+static void _regulator_runaway_error() {
+  quickStop();
+  disable_all_heaters();
+  disable_all_steppers();
+  SERIAL_PROTOCOLLNPGM("Runaway Pressure Error. Canceling print.");
+  SERIAL_PROTOCOLLNPGM("// action:cancel");
+  Regulator__SetOutputPressure(0.0);
```
Comment (pizzyflavin) [inline Marlin/Regulator.cpp]: Same, maybe set this above the SERIAL stuff

File: Marlin/Regulator.cpp [hunk 20]
```diff
@@ -33,4 +69,98 @@ void setOutputPressure(float desired_pressure) {
...
+  sled_present = true;
+}
+
+/** 
+ * Allows us to determine if the pressure is too high or low, indicating
+ * a serious issue. Checks the target temperature vs. the desired, and
```
Comment (pizzyflavin) [inline Marlin/Regulator.cpp]: Target temperature?

File: Marlin/Regulator.cpp [hunk 21]
```diff
@@ -33,4 +69,98 @@ void setOutputPressure(float desired_pressure) {
...
+  Regulator__SetOutputPressure(0.0);
+  sled_present = true;
+}
+
+/** 
+ * Allows us to determine if the pressure is too high or low, indicating
+ * a serious issue. Checks the target temperature vs. the desired, and 
+ * throws an error if outside the bounds for too long.
+ * @target_pressure  The pressure you're attempting to reach
+ * @pressure         The actual current pressure
```
Comment (pizzyflavin) [inline Marlin/Regulator.cpp]: I'm not positive, but I _think_ this is inconsistent with the previous documentation that uses @param and then gives the parameter name and explanantion

File: Marlin/Regulator.cpp [hunk 22]
```diff
@@ -33,4 +69,98 @@ void setOutputPressure(float desired_pressure) {
...
+      if (pressure == REGULATOR_NOT_PRESENT_VALUE) {
+        _regulator_sled_removed_error();
+      } else if (regulatorOver) {
+        _regulator_runaway_error();
+      } else {
+        _regulator_leak_error();
+      }
+      regulatorTimer = millis();
+    }
+  } else {
```
Comment (pizzyflavin) [inline Marlin/Regulator.cpp]: Looks like `regulatorTimer = millis()` is going to happen no matter what, so you could remove the else statement and put `regulatorTimer = millis()` after the if with no else

Comment (kdumontnu) [inline Marlin/Regulator.cpp]: I was confused as well. Please comment and format this a little more clearly.

File: Marlin/Regulator.cpp (hunk 8 above)
Comment (danthompson41) [inline Marlin/Regulator.cpp]: Good call, I pretty much left it there for myself haha but it's not super portable / relevant anymore.

File: Marlin/Regulator.cpp (hunk 9 above)
Comment (danthompson41) [inline Marlin/Regulator.cpp]: This should be implemented, was probably a debug thing that I pulled it out. I'll try it with it back in and make sure that it works as intended.

File: Marlin/Regulator.cpp (hunk 21 above)
Comment (danthompson41) [inline Marlin/Regulator.cpp]: Yep, I used parameter before. Changing it to param in that file as it seems to be more doxygen compliant.

File: Marlin/Marlin_main.cpp (hunk 7 above)
Comment (danthompson41) [inline Marlin/Marlin_main.cpp]: Yep, that's totally just a ram concern

File: Marlin/Regulator.cpp (hunk 22 above)
Comment (danthompson41) [inline Marlin/Regulator.cpp]: Sure. There's an important case where we're not updating it, which is when we're under or over and we haven't waited long enough to hit an error yet. I'll make this prettier

**Comment by danthompson41** [comment]
Got verbal approval, merging in
